### PR TITLE
[->equality-based-deletes-main] Delete Strategy Implementation and Compactor V2 merge algorithm changes

### PR DESCRIPTION
**Pull request by pfaraone**
## Description
- Implementation of equality-based delete strategy and compactor V2 merge step changes

### Linked PRs
- https://github.com/ray-project/deltacat/pull/278
- https://github.com/ray-project/deltacat/pull/273

## Testing
- Additional rebase then incremental test cases added
- prepareDeletes unit testing added

### CODE REVIEW

File: deltacat/compute/compactor_v2/steps/merge.py [hunk 1]
```diff
@@ -257,14 +257,170 @@ def _copy_all_manifest_files_from_old_hash_buckets(
...
+    iter_df_envelopes = iter(df_envelopes)
+    upserts, deletes = [], []
+    for is_upsert, group_envelopes in itertools.groupby(
+        iter_df_envelopes, lambda x: x.delta_type == DeltaType.UPSERT
+    ):
+        if is_upsert:
+            upserts.extend(group_envelopes)
+        else:
+            deletes.extend(group_envelopes)
+        if deletes or not (deletes or upserts):
```
Comment (raghumdani) [inline deltacat/compute/compactor_v2/steps/merge.py:305]: `not (deletes or upserts)` is true when both are empty. I don't understand this.

File: deltacat/compute/compactor_v2/steps/merge.py [hunk 2]
```diff
@@ -257,14 +257,170 @@ def _copy_all_manifest_files_from_old_hash_buckets(
...
+        yield upserts, []
+
+
+def _compact_table_v2(
+    input: MergeInput,
+    dfe_list: List[List[DeltaFileEnvelope]],
+    hb_idx: int,
+    has_deletes: bool,
+) -> Tuple[pa.Table, int, int]:
+    df_envelopes: List[DeltaFileEnvelope] = _sort_df_envelopes(
```
Comment (raghumdani) [inline deltacat/compute/compactor_v2/steps/merge.py:318]: Looks like validation here has gone missing that only UPSERT and DELETE deltas are supported.

File: deltacat/compute/compactor_v2/steps/merge.py [hunk 3]
```diff
@@ -257,14 +257,170 @@ def _copy_all_manifest_files_from_old_hash_buckets(
...
+        )
+        if has_deletes:
+            table, partial_dropped_rows = input.delete_strategy.apply_all_deletes(
+                table, input.delete_file_envelopes
+            )
+            total_dropped_records += partial_dropped_rows
+    prev_table = table
+    for i, (upsert_group, delete_group) in enumerate(
+        group_by_upsert_delete_sequence(all_dfes)
+    ):
```
Comment (raghumdani) [inline deltacat/compute/compactor_v2/steps/merge.py:345]: This is complicated. You don't need to process `upsert_group` and `delete_group` at once. The former requires you to call `_apply_upserts` and latter requires you to call `apply_all_deletes`. At the end of the day, you have an arrow table as result.

File: deltacat/compute/compactor_v2/steps/merge.py [hunk 4]
```diff
@@ -257,14 +257,170 @@ def _copy_all_manifest_files_from_old_hash_buckets(
...
+            rcf=input.round_completion_info,
+            read_kwargs_provider=input.read_kwargs_provider,
+            deltacat_storage=input.deltacat_storage,
+            deltacat_storage_kwargs=input.deltacat_storage_kwargs,
+        )
+        if has_deletes:
+            table, partial_dropped_rows = input.delete_strategy.apply_all_deletes(
+                table, input.delete_file_envelopes
+            )
+            total_dropped_records += partial_dropped_rows
```
Comment (raghumdani) [inline deltacat/compute/compactor_v2/steps/merge.py:341]: Is this necessary to apply all deletes to compacted tables upfront? You will anyway apply when consecutively in the for loop below.

File: deltacat/compute/compactor_v2/steps/merge.py [hunk 5]
```diff
@@ -257,14 +257,170 @@ def _copy_all_manifest_files_from_old_hash_buckets(
...
+            total_dropped_records += dropped_rows
+        prev_table = table
+    return (
+        table,
+        total_incremental_len,
+        total_deduped_records,
+    )
+
+
+def _compact_table_with_previously_compacted_table(
```
Comment (raghumdani) [inline deltacat/compute/compactor_v2/steps/merge.py:378]: Can be renamed to `_apply_upserts`.

File: deltacat/compute/compactor_v2/steps/merge.py [hunk 6]
```diff
@@ -350,18 +506,25 @@ def _timed_merge(input: MergeInput) -> MergeResult:
         hb_index_copy_by_ref_ids = []
 
         for merge_file_group in merge_file_groups:
-            if not merge_file_group.dfe_groups:
+            has_delete = (
+                input.delete_file_envelopes is not None
+                and input.delete_strategy is not None
+            )
```
Comment (raghumdani) [inline deltacat/compute/compactor_v2/steps/merge.py:514]:
```suggestion
            has_delete = input.delete_file_envelopes is not None
            if has_delete:
                assert(input.delete_strategy, "Merge input missing delete_strategy")
```

File: deltacat/compute/compactor_v2/steps/merge.py [hunk 7]
```diff
@@ -257,14 +257,170 @@ def _copy_all_manifest_files_from_old_hash_buckets(
...
+    return (
+        table,
+        total_incremental_len,
+        total_deduped_records,
+    )
+
+
+def _compact_table_with_previously_compacted_table(
+    input: MergeInput,
+    dfe_list: List[DeltaFileEnvelope],
```
Comment (raghumdani) [inline deltacat/compute/compactor_v2/steps/merge.py:380]: Add assertion here that all dfes are UPSERT.

File: deltacat/compute/compactor_v2/steps/merge.py [hunk 8]
```diff
@@ -257,14 +257,170 @@ def _copy_all_manifest_files_from_old_hash_buckets(
...
+    )
+    total_deduped_records = hb_table_record_count - len(table)
+    logger.info(
+        f"[Merge task index {input.merge_task_index}] Merged "
+        f"record count: {len(table)}, size={table.nbytes} took: {merge_time}s"
+    )
+    return table, incremental_len, total_deduped_records
+
+
 def _compact_tables(
```
Comment (raghumdani) [inline deltacat/compute/compactor_v2/steps/merge.py:415]: You can directly modify this method. No need to have `_compact_tables_v2`.

File: deltacat/tests/compute/compact_partition_rebase_then_incremental_test_cases.py [hunk 9]
```diff
@@ -0,0 +1,1347 @@
...
+            names=["pk_col_1", "col_1"],
+        ),
+        do_create_placement_group=True,
+        records_per_compacted_file=DEFAULT_MAX_RECORDS_PER_FILE,
+        hash_bucket_count=DEFAULT_HASH_BUCKET_COUNT,
+        read_kwargs_provider=None,
+        drop_duplicates=True,
+        skip_enabled_compact_partition_drivers=[CompactorVersion.V1],
+    ),
+    "22-rebase-then-incremental-delete-type-delta-UDU-U-affects-compacted-and-incremental": RebaseThenIncrementalCompactionTestCaseParams(
```
Comment (raghumdani) [inline deltacat/tests/compute/compact_partition_rebase_then_incremental_test_cases.py:1278]: Neat!

File: deltacat/compute/compactor_v2/compaction_session.py [hunk 10]
```diff
@@ -24,6 +24,16 @@
 from deltacat.compute.compactor_v2.utils.merge import (
     generate_local_merge_input,
 )
+from deltacat.compute.compactor import DeltaAnnotated
```
Comment (raghumdani) [inline deltacat/compute/compactor_v2/compaction_session.py:27]: Please also change memory estimation for merge.

File: deltacat/compute/compactor_v2/steps/merge.py [hunk 11]
```diff
@@ -257,14 +257,162 @@ def _copy_all_manifest_files_from_old_hash_buckets(
...
+    total_dropped_records = 0
+    if _does_hash_bucket_idx_have_compacted_table(input, hb_idx):
+        table = _download_compacted_table(
+            hb_index=hb_idx,
+            rcf=input.round_completion_info,
+            read_kwargs_provider=input.read_kwargs_provider,
+            deltacat_storage=input.deltacat_storage,
+            deltacat_storage_kwargs=input.deltacat_storage_kwargs,
+        )
+        if has_deletes:
```
[the remark this answers is not in the document]
Comment (pfaraone) [reply deltacat/compute/compactor_v2/steps/merge.py:385]: Yes - this is passed as an argument _compact_table_v2 and will be False if there are no delete_file_envelopes

File: deltacat/compute/compactor_v2/deletes/delete_strategy_equality_delete.py [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+        if not table or not delete_file_envelope.table:
+            return table, 0
+        delete_columns = delete_file_envelope.delete_columns
+        delete_table = delete_file_envelope.table
+        table, number_of_rows_dropped = self._drop_rows(
+            table, delete_table, delete_columns
+        )
+        return table, number_of_rows_dropped
+
+    def apply_all_deletes(
```
Comment (Zyiqin-Miranda) [inline deltacat/compute/compactor_v2/deletes/delete_strategy_equality_delete.py:86]: What's the difference between `apply_deletes` and `apply_all_deletes`?

File: deltacat/compute/compactor_v2/deletes/delete_file_envelope.py [hunk 13]
```diff
@@ -56,8 +56,13 @@ def of(
         )
         assert len(delete_columns) > 0, "At least 1 delete column is expected"
         delete_file_envelope["delete_columns"] = delete_columns
+        delete_file_envelope["table_size_bytes"] = table.nbytes
```
Comment (raghumdani) [inline deltacat/compute/compactor_v2/deletes/delete_file_envelope.py:59]: This assumes that table is a pyarrow table but not LocalTable.

File: deltacat/compute/compactor_v2/steps/merge.py [hunk 14]
```diff
@@ -257,62 +251,219 @@ def _copy_all_manifest_files_from_old_hash_buckets(
...
+            table, partial_dropped_rows = input.delete_strategy.apply_all_deletes(
+                table, input.delete_file_envelopes
+            )
+            aggregated_dropped_records += partial_dropped_rows
+    prev_table = table
+    for i, (upsert_group, delete_group) in enumerate(
+        _group_by_upsert_delete_sequence(all_dfes)
+    ):
```
Comment (raghumdani) [inline deltacat/compute/compactor_v2/steps/merge.py:394]: Should we process one group at once? When processing upsert we call apply_upersts and during deletes we call apply_deletes?

File: deltacat/compute/compactor/model/table_object_store.py [hunk 15]
```diff
@@ -22,6 +22,9 @@ def store_table(self, table: LocalTable) -> LocalTableReference:
     def get_table(self, table_like: LocalTableReference) -> LocalTable:
         pass
 
+    @abstractmethod
+    def get_table_reference(self, table_ref: Any) -> Any:
```
Comment (raghumdani) [inline deltacat/compute/compactor/model/table_object_store.py:26]: Passing a `table_ref` and getting the `table_reference` back? This method doesn't seem to be doing anything.

File: deltacat/tests/compute/compact_partition_rebase_then_incremental_test_cases.py [hunk 16]
```diff
@@ -0,0 +1,1347 @@
...
+                        pa.array([2, 3, 6, 7]),
+                        pa.array(["boo", "bar", "fiz", "aaa"]),
+                    ],
+                    names=["pk_col_1", "col_1"],
+                ),
+                DeltaType.UPSERT,
+                None,
+            ),
+            (
+                pa.Table.from_arrays(
```
Comment (raghumdani) [inline deltacat/tests/compute/compact_partition_rebase_then_incremental_test_cases.py:1254]: Can you add a test case with an empty delete i.e., without any `delete_columns`? and with delete column but no records?

File: deltacat/compute/compactor_v2/deletes/delete_strategy_equality_delete.py [hunk 17]
```diff
@@ -0,0 +1,117 @@
...
+                table[delete_column_name],
+                value_set=delete_table[delete_column_name],
+            )
+            result = equality_predicate_operation(prev_boolean_mask, curr_boolean_mask)
+            prev_boolean_mask = result
+        number_of_rows_before_dropping = len(table)
+        logger.debug(
+            f"Number of table rows before dropping: {number_of_rows_before_dropping}."
+        )
+        table = table.filter(pc.invert(result))
```
Comment (raghumdani) [inline deltacat/compute/compactor_v2/deletes/delete_strategy_equality_delete.py:68]: It looks like you will drop everything if `delete_column_names` is empty.

File: deltacat/compute/compactor_v2/deletes/delete_strategy_equality_delete.py [hunk 18]
```diff
@@ -0,0 +1,117 @@
...
+
+        Returns:
+            Tuple[pa.Table, int]: A tuple containing the updated pyarrow table after applying all deletes,
+                and the total number of rows deleted.
+        """
+        # looking up table references are lighter than the actual table
+        if not table or not all(
+            delete_file_envelope.table_reference is not None
+            for delete_file_envelope in delete_file_envelopes
+        ):
```
Comment (raghumdani) [inline deltacat/compute/compactor_v2/deletes/delete_strategy_equality_delete.py:125]: The latter condition doesn't look right: `not all(True, False, False)` will yield `True` but we want to drop rows if atleast one dfe has deletes.

File: deltacat/compute/compactor_v2/deletes/utils.py [hunk 19]
```diff
@@ -0,0 +1,153 @@
...
+
+def _filter_out_non_delete_deltas(input_deltas: List[Delta]) -> List[Delta]:
+    non_delete_deltas = []
+    for input_delta in input_deltas:
+        if input_delta.type is not DeltaType.DELETE:
+            non_delete_deltas.append(input_delta)
+    return non_delete_deltas
+
+
+def _aggregate_delete_deltas(input_deltas: List[Delta]) -> Dict[int, List[Delta]]:
```
Comment (raghumdani) [inline deltacat/compute/compactor_v2/deletes/utils.py:34]: This method is still not changed.

File: deltacat/compute/compactor_v2/steps/merge.py [hunk 20]
```diff
@@ -345,23 +483,44 @@ def _timed_merge(input: MergeInput) -> MergeResult:
...
+                has_delete,
+                compacted_table,
             )
             total_input_records += input_records
             total_deduped_records += deduped_records
-            materialized_results.append(
-                merge_utils.materialize(input, merge_file_group.hb_index, [table])
-            )
+            total_dropped_records += dropped_records
+            if table is not None:
```
Comment (raghumdani) [inline deltacat/compute/compactor_v2/steps/merge.py:520]: will table ever be None at this point?

**Review by raghumdani — COMMENTED** [review]
Most comments are already addressed and is now in great shape.

File: deltacat/compute/compactor_v2/deletes/utils.py [hunk 21]
```diff
@@ -0,0 +1,142 @@
...
+            window_end += 1
+        delete_deltas_sequence: List[Delta] = input_deltas[window_start:window_end]
+        stream_position_of_earliest_delete_in_sequence: int = delete_deltas_sequence[
+            0
+        ].stream_position
+        delete_delta_sequence_spos_to_delete_delta[
+            stream_position_of_earliest_delete_in_sequence
+        ].extend(delete_deltas_sequence)
+        window_start = window_end
+    return delete_delta_sequence_spos_to_delete_delta
```
[the remark this answers is not in the document]
Comment (raghumdani) [reply deltacat/compute/compactor_v2/deletes/utils.py]: No strong bias but this refactoring should make it much simpler to understand.

File: deltacat/compute/compactor_v2/steps/merge.py (hunk 14 above)
Comment (pfaraone) [reply deltacat/compute/compactor_v2/steps/merge.py:381]: I don't follow.  `_group_by_upsert_delete_sequence(all_dfes)` is an iterator that yields a sequence of Upserts bookended by Deletes if present. Once we have the list of upsert and deletes we are operating on list of upserts and deletes DeltaFileEnvelopes (not individual elements). How would you refactor it? 

```
for i, (upsert_group, delete_group) in enumerate(
        _group_by_upsert_delete_sequence(all_dfes)
    ):
        current_upsert_dfe_sequence: List[DeltaFileEnvelope] = list(upsert_group)
        current_delete_dfe_sequence: List[DeltaFileEnvelope] = list(delete_group)
         if current_upsert_dfe_sequence:
               _apply_upserts()
         if current_delete_dfe_sequence:
              input.delete_strategy.apply_many_deletes()
```

File: deltacat/compute/compactor/model/table_object_store.py (hunk 15 above)
Comment (pfaraone) [reply deltacat/compute/compactor/model/table_object_store.py]: Removed this abstract method

File: deltacat/compute/compactor_v2/deletes/utils.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (pfaraone) [reply deltacat/compute/compactor_v2/deletes/utils.py]: I tried unit testing the suggested change. The suggested change skips over non DELETE deltas properly but doesn't handle sequential deletes with different parameters (which should be added as different entries to start_stream_spos_to_delete_deltas). I'd prefer to stick with the sliding window approach

[the remark this answers is not in the document]
Comment (raghumdani) [reply deltacat/compute/compactor_v2/deletes/utils.py]: Just add that check `current_list[0].delete_parameters == current.delete_parameters` and it should work.

File: deltacat/compute/compactor_v2/steps/merge.py (hunk 14 above)
Comment (raghumdani) [reply deltacat/compute/compactor_v2/steps/merge.py:381]: Do you have to group upserts with deletes? You could iterate on group of upserts and deletes separately.

File: deltacat/compute/compactor_v2/deletes/utils.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (pfaraone) [reply deltacat/compute/compactor_v2/deletes/utils.py]: Changed - please review it again

File: deltacat/compute/compactor_v2/deletes/utils.py (hunk 19 above)
Comment (pfaraone) [reply deltacat/compute/compactor_v2/deletes/utils.py:28]: Changed now

```
def _aggregate_delete_deltas(input_deltas: List[Delta]) -> Dict[int, List[Delta]]:
    """
    Aggregates consecutive DELETE deltas with the same delete parameters into groups.

    Args:
        input_deltas (List[Delta]): A list of Delta objects representing delete operations.
    Returns:
        Dict[int, List[Delta]]: A dictionary where the keys are the stream positions of the
        earliest delta in each group of consecutive DELETE deltas with the same delete parameters,
        and the values are lists containing those deltas.
    """
    start_stream_spos_to_delete_delta_sequence: Dict[int, List[Delta]] = defaultdict(
        list
    )
    delete_deltas_sequence_grouped_by_delete_parameters: List[
        Tuple[bool, List[Delta]]
    ] = [
        (is_delete, sorted(list(delete_delta_group), key=lambda x: x.stream_position))
        for (is_delete, _), delete_delta_group in itertools.groupby(
            input_deltas, lambda d: (d.type is DeltaType.DELETE, d.delete_parameters)
        )
    ]
    for (
        is_delete_delta_sequence,
        delete_delta_sequence,
    ) in delete_deltas_sequence_grouped_by_delete_parameters:
        if not is_delete_delta_sequence:
            continue
        starting_stream_position_of_delete_sequence: int = delete_delta_sequence[
            0
        ].stream_position
        start_stream_spos_to_delete_delta_sequence[
            starting_stream_position_of_delete_sequence
        ] = delete_delta_sequence
    return start_stream_spos_to_delete_delta_sequence
```

File: deltacat/compute/compactor_v2/steps/merge.py (hunk 14 above)
Comment (pfaraone) [reply deltacat/compute/compactor_v2/steps/merge.py:381]: Changed
```
def _compact_table_v2(
    input: MergeInput,
    dfe_list: List[List[DeltaFileEnvelope]],
    hb_idx: int,
    has_deletes: bool,
    compacted_table: Optional[pa.Table] = None,
) -> Tuple[pa.Table, int, int, int]:
    """
    Compacts a list of DeltaFileEnvelope objects into a single PyArrow table.

    Args:
        input (MergeInput): The input for the merge operation.
        dfe_list (List[List[DeltaFileEnvelope]]): A list of lists of DeltaFileEnvelope objects.
        hb_idx (int): The hash bucket index for the compaction.
        has_deletes (bool): Whether the input data contains DELETE operations.

    Returns:
        Tuple[pa.Table, int, int, int]: A tuple containing:
            1. The compacted PyArrow table.
            2. The total number of records in the incremental data.
            3. The total number of deduplicated records.
            4. The total number of dropped records due to DELETE operations.
    """
    df_envelopes: List[DeltaFileEnvelope] = _flatten_dfe_list(dfe_list)
    delete_file_envelopes = input.delete_file_envelopes or []
    reordered_all_dfes: List[DeltaFileEnvelope] = _sort_df_envelopes(
        delete_file_envelopes + df_envelopes
    )
    assert all(
        dfe.delta_type in (DeltaType.UPSERT, DeltaType.DELETE) for dfe in reordered_all_dfes
    ), "All reordered delta file envelopes must of the UPSERT or DELETE"
    prev_table, table = compacted_table, compacted_table
    aggregated_incremental_len = 0
    aggregated_deduped_records = 0
    aggregated_dropped_records = 0
    for i, (delta_type, delta_type_sequence) in enumerate(_group_by_sequence(reordered_all_dfes)):
        if delta_type is DeltaType.UPSERT:
            (
                table,
                partial_incremental_len,
                partial_deduped_records,
                partial_merge_time,
            ) = _apply_upserts(input, delta_type_sequence, hb_idx, prev_table)
            logger.info(
                f"[Merge task index {input.merge_task_index}] Merged "
                f"record count: {len(table)}, size={table.nbytes} took: {partial_merge_time}s"
            )
            aggregated_incremental_len += partial_incremental_len
            aggregated_deduped_records += partial_deduped_records
        elif delta_type is DeltaType.DELETE:
            (table, dropped_rows) = input.delete_strategy.apply_many_deletes(
                table, delta_type_sequence
            )
            logger.info(
                f"[Merge task index {input.merge_task_index}] Dropped "
                + f"record count: {dropped_rows}"
            )
            aggregated_dropped_records += dropped_rows
        prev_table = table
    return (
        table,
        aggregated_incremental_len,
        aggregated_deduped_records,
        aggregated_dropped_records,
    )
```

File: deltacat/compute/compactor_v2/steps/merge.py (hunk 20 above)
Comment (pfaraone) [reply deltacat/compute/compactor_v2/steps/merge.py:504]: Fixed now - the resulting table would be None if `merge_file_group.dfe_groups` is `None` and `compacted_table` is `None`

File: deltacat/compute/compactor_v2/deletes/utils.py [hunk 22]
```diff
@@ -0,0 +1,157 @@
...
+    assert all(
+        input_deltas[i].stream_position <= input_deltas[i + 1].stream_position
+        for i in range(len(input_deltas) - 1)
+    ), "Uniform deltas must be in non-decreasing order by stream position"
+    start_stream_spos_to_delete_delta_sequence: Dict[
+        int, List[Delta]
+    ] = _aggregate_delete_deltas(input_deltas)
+    delete_file_envelopes: List[DeleteFileEnvelope] = _get_delete_file_envelopes(
+        params, start_stream_spos_to_delete_delta_sequence
+    )
```
Comment (raghumdani) [inline deltacat/compute/compactor_v2/deletes/utils.py:153]: As a safety measure, assert the length of `start_stream_spos_to_delete_delta_sequence` and `delete_file_envelopes` to be equal.

File: deltacat/compute/compactor_v2/steps/merge.py [hunk 23]
```diff
@@ -69,17 +70,12 @@ def _drop_delta_type_rows(table: pa.Table, delta_type: DeltaType) -> pa.Table:
 
 
 def _build_incremental_table(
-    df_envelopes_list: List[List[DeltaFileEnvelope]],
+    sorted_df_envelopes: List[DeltaFileEnvelope],
```
Comment (raghumdani) [inline deltacat/compute/compactor_v2/steps/merge.py:73]: just rename this to `df_envelopes`.

File: deltacat/compute/compactor_v2/steps/merge.py [hunk 24]
```diff
@@ -257,62 +251,190 @@ def _copy_all_manifest_files_from_old_hash_buckets(
...
+    """
+    df_envelopes: List[DeltaFileEnvelope] = _flatten_dfe_list(dfe_list)
+    delete_file_envelopes = input.delete_file_envelopes or []
+    reordered_all_dfes: List[DeltaFileEnvelope] = _sort_df_envelopes(
+        delete_file_envelopes + df_envelopes
+    )
+    assert all(
+        dfe.delta_type in (DeltaType.UPSERT, DeltaType.DELETE)
+        for dfe in reordered_all_dfes
+    ), "All reordered delta file envelopes must of the UPSERT or DELETE"
```
Comment (raghumdani) [inline deltacat/compute/compactor_v2/steps/merge.py:363]:
```suggestion
    ), "All reordered delta file envelopes must be of the UPSERT or DELETE"
```

File: deltacat/compute/compactor_v2/compaction_session.py [hunk 25]
```diff
@@ -507,26 +539,24 @@ def merge_input_provider(index, item):
         and compaction_audit.hash_bucket_processed_size_bytes
     ):
         input_inflation = (
-            compaction_audit.hash_bucket_processed_size_bytes
-            / compaction_audit.input_size_bytes
-        )
+            compaction_audit.hash_bucket_processed_size_bytes + delete_file_size_bytes
```
Comment (raghumdani) [inline deltacat/compute/compactor_v2/compaction_session.py:542]: Why is this changed? We need to change in `task_options.py` to account for merge memory estimation.

File: deltacat/tests/compute/compact_partition_rebase_then_incremental_test_cases.py [hunk 26]
```diff
@@ -0,0 +1,1533 @@
...
+                    ],
+                    schema=pa.schema(
+                        [
+                            ("pk_col_1", pa.int64()),
+                            ("col_1", pa.string()),
+                        ]
+                    ),
+                ),
+                DeltaType.DELETE,
+                None,
```
Comment (raghumdani) [inline deltacat/tests/compute/compact_partition_rebase_then_incremental_test_cases.py:1503]: When delete column is present but not records should succeed without changing the table.

File: deltacat/tests/compute/compact_partition_rebase_then_incremental_test_cases.py [hunk 27]
```diff
@@ -0,0 +1,1533 @@
...
+        ),
+        expected_terminal_exception=AssertionError,
+        do_create_placement_group=True,
+        records_per_compacted_file=DEFAULT_MAX_RECORDS_PER_FILE,
+        hash_bucket_count=DEFAULT_HASH_BUCKET_COUNT,
+        read_kwargs_provider=None,
+        drop_duplicates=True,
+        skip_enabled_compact_partition_drivers=[CompactorVersion.V1],
+    ),
+    "24-rebase-then-incremental-delete-type-delta-delete-column-no-records": RebaseThenIncrementalCompactionTestCaseParams(
```
Comment (raghumdani) [inline deltacat/tests/compute/compact_partition_rebase_then_incremental_test_cases.py:1457]: Another test case where there are duplicate delete records in a DELETE delta.

**Review by raghumdani — COMMENTED** [review]
LGTM! Please address minor comments.

File: deltacat/compute/compactor_v2/deletes/delete_strategy_equality_delete.py (hunk 18 above)
Comment (pfaraone) [reply deltacat/compute/compactor_v2/deletes/delete_strategy_equality_delete.py]: Fixed 

```
def apply_many_deletes(
        self,
        table: Optional[pa.Table],
        delete_file_envelopes: List[DeleteFileEnvelope],
        *args,
        **kwargs,
    ) -> Tuple[pa.Table, int]:
        """
        Apply delete operations on the given table using all provided delete file envelopes.

        Args:
            table (Optional[pa.Table]): The pyarrow table to apply deletes on.
            delete_file_envelopes (List[DeleteFileEnvelope]): A list of delete file envelopes containing delete parameters.

        Returns:
            Tuple[pa.Table, int]: A tuple containing the updated pyarrow table after applying all deletes,
                and the total number of rows deleted.
        """
        # looking up table references are lighter than the actual table
        if not table or any(
            delete_file_envelope.table_reference is None
            for delete_file_envelope in delete_file_envelopes
        ):
            logger.warning(
                "Delete file envelopes are missing a delete table. Not applying any deletes"
            )
            return table, 0
```

File: deltacat/compute/compactor_v2/deletes/delete_strategy_equality_delete.py [hunk 28]
```diff
@@ -0,0 +1,143 @@
...
+    def _drop_rows(
+        self,
+        table: pa.Table,
+        delete_table: pa.Table,
+        delete_column_names: List[str],
+        equality_predicate_operation: Optional[Callable] = pa.compute.and_,
+    ) -> Tuple[pa.Table, int]:
+        if len(delete_column_names) < 1:
+            return table, 0
+        prev_boolean_mask = pa.array(np.zeros(len(table), dtype=bool))
```
[the remark this answers is not in the document]
Comment (pfaraone) [reply deltacat/compute/compactor_v2/deletes/delete_strategy_equality_delete.py]: Added explanatory comment

File: deltacat/compute/compactor_v2/deletes/utils.py [hunk 29]
```diff
@@ -0,0 +1,160 @@
...
+                **params.deltacat_storage_kwargs,
+            )
+            consecutive_delete_tables.extend(delete_dataset)
+        delete_table: pa.Table = pa.concat_tables(consecutive_delete_tables)
+        delete_file_envelope: DeleteFileEnvelope = DeleteFileEnvelope.of(
+            start_stream_position,
+            delta_type=DeltaType.DELETE,
+            table=delete_table,
+            delete_columns=delete_columns,
+        )
```
Comment (rkenmi) [inline deltacat/compute/compactor_v2/deletes/utils.py:109]: Would we also need a way to override `table_storage_strategy` based on what is specified in `params`?

File: deltacat/compute/compactor_v2/steps/merge.py [hunk 30]
```diff
@@ -257,62 +250,190 @@ def _copy_all_manifest_files_from_old_hash_buckets(
...
+        and a list of DeltaFileEnvelope objects that share the same delta_type.
+    """
+    iter_df_envelopes = iter(df_envelopes)
+    for delta_type, delta_type_sequence in itertools.groupby(
+        iter_df_envelopes, lambda x: x.delta_type
+    ):
+        yield delta_type, list(delta_type_sequence)
+
+
+def _compact_table_v2(
```
Comment (rkenmi) [inline deltacat/compute/compactor_v2/steps/merge.py:333]: Minor: I think the naming can be misleading since this module is already under the `compactor_v2` package.

File: deltacat/compute/compactor_v2/steps/merge.py [hunk 31]
```diff
@@ -345,20 +466,44 @@ def _timed_merge(input: MergeInput) -> MergeResult:
...
         total_input_records, total_deduped_records = 0, 0
+        total_dropped_records = 0
         materialized_results: List[MaterializeResult] = []
         merge_file_groups = input.merge_file_groups_provider.create()
         hb_index_copy_by_ref_ids = []
 
         for merge_file_group in merge_file_groups:
-            if not merge_file_group.dfe_groups:
+            compacted_table = None
+            has_delete = input.delete_file_envelopes is not None
```
Comment (rkenmi) [inline deltacat/compute/compactor_v2/steps/merge.py:476]: Have you considered adding delete envelopes to `MergeFileGroup` and updating the provider so that it doesn't have to be directly fed from the input object?

File: deltacat/compute/compactor_v2/deletes/delete_file_envelope.py [hunk 32]
```diff
@@ -56,8 +55,29 @@ def of(
...
+
     @property
     def delete_columns(self) -> List[str]:
         return self["delete_columns"]
+
+    @property
+    def table_reference(self) -> Optional[Any]:
+        if self.table_storage_strategy is not None and isinstance(
+            self.table_storage_strategy,
+            LocalTableRayObjectStoreReferenceStorageStrategy,
```
Comment (rkenmi) [inline deltacat/compute/compactor_v2/deletes/delete_file_envelope.py:80]: Are other implementations a TODO?

File: deltacat/compute/compactor_v2/compaction_session.py [hunk 33]
```diff
@@ -369,6 +369,7 @@ def hash_bucket_input_provider(index, item):
         )
 
         # BSP Step 2: Merge
+        # NOTE: DELETE-type deltas are stored in Plasma object store in prepare_deletes and therefore don't need to included in merge task resource estimation
```
Comment (raghumdani) [inline deltacat/compute/compactor_v2/compaction_session.py:372]:
```suggestion
        # NOTE: DELETE-type deltas are stored in Plasma object store 
        # in prepare_deletes and therefore don't need to included 
        # in merge task resource estimation
```

File: deltacat/compute/compactor_v2/deletes/delete_strategy_equality_delete.py [hunk 34]
```diff
@@ -0,0 +1,163 @@
...
+        Args:
+            table (Optional[pa.Table]): The pyarrow table to apply deletes on.
+            delete_file_envelope (DeleteFileEnvelope): The delete file envelope containing delete parameters.
+
+        Returns:
+            Tuple[pa.Table, int]: A tuple containing the updated Arrow table after applying deletes,
+                and the number of rows deleted.
+        """
+        if not table or not delete_file_envelope.table:
+            return table, 0
```
Comment (rkenmi) [inline deltacat/compute/compactor_v2/deletes/delete_strategy_equality_delete.py:120]: Minor: A debugger statement would be helpful for subtle no-ops like this

File: deltacat/tests/compute/compact_partition_rebase_then_incremental_test_cases.py [hunk 35]
```diff
@@ -0,0 +1,1655 @@
...
+        do_create_placement_group=False,
+        records_per_compacted_file=DEFAULT_MAX_RECORDS_PER_FILE,
+        hash_bucket_count=1,
+        read_kwargs_provider=None,
+        drop_duplicates=True,
+        skip_enabled_compact_partition_drivers=None,
+    ),
+}
+
+REBASE_THEN_INCREMENTAL_DELETE_DELTA_TYPE_TEST_CASES = {
```
Comment (raghumdani) [inline deltacat/tests/compute/compact_partition_rebase_then_incremental_test_cases.py:774]: Can you add a test where DELETE comes first and then UPSERT?

**Review by raghumdani — APPROVED** [review]
LGTM!

File: deltacat/compute/compactor_v2/deletes/delete_strategy_equality_delete.py [hunk 36]
```diff
@@ -0,0 +1,163 @@
...
+        logger.debug(
+            f"Number of table rows after dropping: {number_of_rows_after_dropping}."
+        )
+        dropped_rows = number_of_rows_before_dropping - number_of_rows_after_dropping
+        return table, dropped_rows
+
+    def apply_deletes(
+        self,
+        table: Optional[pa.Table],
+        delete_file_envelope: DeleteFileEnvelope,
```
Comment (rkenmi) [inline deltacat/compute/compactor_v2/deletes/delete_strategy_equality_delete.py:104]: Minor: If this method takes in a generic table to delete and columns to target, then I feel `apply_many_deletes` could easily re-use this  after concatenating all the tables.

**Review by rkenmi — COMMENTED** [review]
Just minor comments/questions, but LGTM!

**Review by Zyiqin-Miranda — APPROVED** [review]
LGTM!

**Review by rkenmi — APPROVED** [review]

File: deltacat/compute/compactor_v2/deletes/utils.py [hunk 37]
```diff
@@ -0,0 +1,157 @@
...
+
+logger = logs.configure_deltacat_logger(logging.getLogger(__name__))
+
+
+def _aggregate_delete_deltas(input_deltas: List[Delta]) -> Dict[int, List[Delta]]:
+    """
+    Aggregates consecutive DELETE deltas with the same delete parameters into groups.
+
+    Args:
+        input_deltas (List[Delta]): A list of Delta objects representing delete operations.
```
[the remark this answers is not in the document]
Comment (pfaraone) [reply deltacat/compute/compactor_v2/deletes/utils.py:33]: `prepare_deletes` needs to be aware of all deltas (upsert and delete) in order to the find sequences of deletes by stream position.

**Review by Zyiqin-Miranda — APPROVED** [review]
LGTM!
